Build performance improvements

fturmel opened the pull request:
Hi @shiffman & team!

I was watching the [stream on YouTube](https://youtu.be/EzYDNydDUfc?t=2391) last Friday and was intrigued by the Gatsby/GraphQL/Netlify log warnings and slow builds. Thought it would be a fun challenge to learn a bit more about these and see if I could help out.

The slowness comes from building the tracks and challenges pages with all filter permutations and pagination. It ends up being a lot of pages!

Here's what I did:

- **Make some of the async code run serially**. There are just too many things running at once, and it ends up making performance worst. This was enough to get rid of the flood of warnings in the logs.
- **Rewrote some GraphQL queries** to make them a bit faster
- **Build contextual languages and topics lists to reduce total permutations and improve UX**. While the challenges and tracks share some topics and languages, there's a lot that doesn't overlap here. This means lots of useless result pages were being generated, and dropdown filters had a lot of options that would return no matches. This appears to have cut the amount of generated pages by more than half.
- **Re-architect how tracks and challenges filters are populated**. We are now generating two static JSON files that are fetched at runtime, which has a lot of positive repercussions:
  - less work for `createPages` process
  - smaller `page-data.json` for every page with filters, and the extracted data found in those new JSON files can be cached by the browser during a user session
  - if there is a topic or language content change, I think that only the relevant pages would get rebuilt instead of every single page that has a filter dropdown component

Some stats on my M1 MacBook Air:

|| total pages built | `createPages` execution |
| --- | ---: | ---: |
|baseline| 19,911 | 51 seconds |
|with this PR| 7,213 | 9 seconds |

**Bonus!**
- Added the `gatsby-plugin-netlify` plugin dependency. It seems to log slow queries during build, so potentially some good information to optimize things further.
- Fixed a few typos in the content, mostly to observe the Netlify build behavior
- Fixed a caching issue. Changes to topics and languages inside track videos and chapters did not bubble up correctly to the parent track level until the Gatsby cache expired or was explicitly purged. This was fixed by using GraphQL resolvers. It should improve the cache deployment issues mentioned in https://github.com/CodingTrain/thecodingtrain.com/pull/786#issuecomment-1322058585

Lastly, I noticed that the Gatsby Netlify plugin is out of date. Maybe that can further help with build speed and caching? I think it's called `Essential Gatsby` in the dashboard. It needs to be uninstalled and re-installed to upgrade https://app.netlify.com/plugins

<img width="1002" alt="Screenshot 2022-11-19 at 1 36 52 PM" src="https://user-images.githubusercontent.com/4009209/202866363-f9e0a513-37fd-488e-a700-3508b108d5dc.png">

Hope that all made sense and I didn't break anything in the process!

Let me know if I can help with something else. Perhaps wrapping up #489 since it would also help reduce the overall amount of built pages by removing duplicated topics and languages?

shiffman commented:
Wow, this is incredible! Everything you are saying here makes sense on a high level to me and the code looks great! Before I merge let's see if @runemadsen, co-founder of DSI, might be able to review as well. #489 is absolutely in need of some attention if you are interested in working on it! 🚂💫

fturmel commented:
Great! I'm reading more on Gatsby (first time using it), and I am realizing that passing the list of topics and languages to the page context will create a bit of client-side duplication because this information will live both in the HTML markup and in JSON for React hydration. It's not a lot of data, but maybe there's a better approach there.

fturmel commented:
I saw a solid improvement locally with these last commits (`createPages` down to 11 seconds), but it's not making a dramatic difference during the Netlify builds 🤷

I think it's worth keeping, but we can also drop these changes. LMK!

shiffman commented:
Hi @fturmel I see this is still active with additional commits, could you let us know when this is ready to review? Thank you!! 💫

fturmel commented:
Sorry for all the noise here. I had an idea this morning I wanted to try out, and it spiraled into trying to troubleshoot incremental builds and cache related issues on the site. I still have a few things I want to look at, I'll let you know when it's ready and will keep it in draft mode until then. Hope that's alright!

fturmel commented:
@shiffman Thanks for your patience, I think we're in a great spot now and there's nothing more for me to change. I rebased and squashed all the noise and amended the initial PR description with additional info.

shiffman commented:
@fturmel wonderful, thank you so much for spending so much time on this!! I'll be glad to merge once @runemadsen is able to take a look! 🚀

fturmel commented:
@runemadsen Great, thanks for reviewing!

> the reason why the Netlify builds take longer is because the Netlify build servers don't have a lot of CPU [...]

For sure, weaker CPUs although I think they have 16 cores. The CI/CD also has more bookend work to do (load caches, setup, deploy to CDN, etc). When I wrote that comment, I was pondering about a specific commit during the PR process (squashed), I think related to the GraphQL regex query change. I saw something like an 8 seconds reduction for `createPages` locally, but it didn't move the needle at all during the Netlify build. Maybe it's a CPU architecture thing...

And to clarify, this PR _does_ improve the Netlify build and deploy numbers. They appear to be hovering between 6 and 8 minutes, with one (fluke?) even completing in 4 minutes. Most current builds are in the 10 to 12+ minutes range.

Revising this PR again this morning, and as one last step here we could remove the `Tag` nodes altogether. They aren't queried anymore and have been proven to become stale. It would be better to get rid of them to prevent their use in upcoming feature implementations. If that makes sense, I can push one more commit to wrap this up. Have a great day!

https://github.com/CodingTrain/thecodingtrain.com/blob/a9f9226f70cabc4a5014f817c92f2ac9af4d91a0/node-scripts/schema.js#L251-L254

https://github.com/CodingTrain/thecodingtrain.com/blob/a9f9226f70cabc4a5014f817c92f2ac9af4d91a0/node-scripts/node-generation.js#L144-L172

fturmel commented:
@shiffman Do you want to try to upgrade the Netlify plugin in the dashboard as well? It's the "Essential Gatsby" one https://app.netlify.com/plugins

<img width="994" alt="Screenshot 2022-11-23 at 7 43 50 AM" src="https://user-images.githubusercontent.com/4009209/203549917-635e62f8-9054-4b4e-bed2-10df1b6f7559.png">

shiffman commented:
Strange, according to the settings I am on the latest release?

<img width="954" alt="Screen Shot 2022-11-23 at 9 35 28 AM" src="https://user-images.githubusercontent.com/191758/203573544-60d89260-607f-4294-aaee-5f45606037ed.png">

Am I missing another place to look?

fturmel commented:
Hrm.. I think this screen only displays the latest version number available and not necessarily the version installed. Plugins are installed at the project level, so you probably need to drill down a bit further. See what's available in the options dropdown.

shiffman commented:
Oh indeed, I was looking in the wrong place it's quite obvious, upgrading now!

runemadsen commented:
@fturmel I'm all for removing the `Tag` type if it's not being used!

fturmel commented:
Alright, good to go!

shiffman commented:
And merging! 🚀
